.github/cloudbuild/android.yaml

```diff
@@ -0,0 +1,111 @@
+  - -xc
+  - |
+    DOCKERFILE_COMMIT=$$(git log -n 1 --pretty=format:%H -- android/docker/mlperf_mobile/Dockerfile)
+    docker build \
+      -t gcr.io/$PROJECT_ID/mlperf-mobile-android:$$DOCKERFILE_COMMIT \
+      -t gcr.io/$PROJECT_ID/mlperf-mobile-android:latest \
+      --cache-from gcr.io/$PROJECT_ID/mlperf-mobile-android:$$DOCKERFILE_COMMIT \
+      --cache-from gcr.io/$PROJECT_ID/mlperf-mobile-android:latest \
+      -f android/docker/mlperf_mobile/Dockerfile \
+      .
+# If the build fails artifacts are not uploaded automatically, so we save them manually before build
+- id: push-new-image
+  name: gcr.io/cloud-builders/docker
+  timeout: 600s # 10 minutes
+  entrypoint: bash
+  args:
+  - -xc
+  - |
+    DOCKERFILE_COMMIT=$$(git log -n 1 --pretty=format:%H -- android/docker/mlperf_mobile/Dockerfile)
+    docker push gcr.io/$PROJECT_ID/mlperf-mobile-android:$$DOCKERFILE_COMMIT
+    docker push gcr.io/$PROJECT_ID/mlperf-mobile-android:latest
+- id: build-apks
+  name: gcr.io/$PROJECT_ID/mlperf-mobile-android:latest
+  waitFor:
+  - build-new-image # we can run this step in parallel with image uploading
+  timeout: 10800s # 3 hours
+  entrypoint: bash
+  args:
+  - -xc
+  - |
+    bazel-3.7.2 build \
+      --remote_cache=https://storage.googleapis.com/$_BAZEL_CACHE_BUCKET \
+      --google_default_credentials \
+      -c opt \
+      --config android_arm64 \
+      //android/java/org/mlperf/inference:mlperf_app //android/androidTest:mlperf_test_app
+    # artifacts can not be uploaded from a symlinked directory, like bazel-bin
+    mkdir -p output/artifacts
+    cp bazel-bin/android/java/org/mlperf/inference/mlperf_app.apk output/artifacts
+    cp bazel-bin/android/androidTest/mlperf_test_app.apk output/artifacts
+- id: instrument-test-on-firebase
+  name: gcr.io/cloud-builders/gcloud
+  timeout: 3600s # 1 hour
+  entrypoint: bash
+  args:
+  - -xc
+  - |
+    # redfin is Pixel 5e
+    gcloud firebase test android run \
+      --type instrumentation \
+      --app output/artifacts/mlperf_app.apk \
+      --test output/artifacts/mlperf_test_app.apk \
+      --timeout 45m \
+      --device model=redfin,version=30,locale=en,orientation=portrait
+- id: crawler-test-on-firebase
+  name: gcr.io/cloud-builders/gcloud
+  waitFor:
+  - build-apks # don't wait for the other test to finish, run as soon as APKs are ready
+  timeout: 3600s # 1 hour (we sometimes wait 30+ minutes until the device we need is free)
+  entrypoint: bash
+  args:
+  - -xc
+  - |
+    # x1q is SM-G981U1 (Samsung Galaxy S20 5G)
+    gcloud firebase test android run \
+      --type robo \
+      --app output/artifacts/mlperf_app.apk \
+      --device model=x1q,version=29,locale=en,orientation=portrait \
+      --timeout 90s
+
+# We uploaded both tags earlier, but this option also adds them to the artifacts page of the build
+images:
+- gcr.io/$PROJECT_ID/mlperf-mobile-android
+
+artifacts:
+  objects:
+    location: gs://$_ARTIFACT_BUCKET/$_ARTIFACT_FOLDER/$COMMIT_SHA

```

Comment:
I actually used to browser through the file system to download a certain artifact because I did not have the artifact tab :)
But you are right, in this case just using the long hash is ok.